### PR TITLE
Added detail to load path documentation

### PR DESCRIPTION
**Pull request by TikhonJelvis**
Two changes:

  1. Added details about the Theta load path to the user guide

  2. Updated and expanded code documentation for `loadModule` TH function

@dmvianna Does this provide a good answer to #10? Thanks for asking that, by the way—without questions like that, I would not know where to improve the documentation, and that's a much bigger bottleneck than the writing itself.

### CODE REVIEW

**Comment by pasunboneleve** [comment]
While you were making the changes I eventually figured out what to do. I believe the documentation confuses _namespaces_ and _modules_ with _directories_ and _filenames_. The way I understand it, modules would be 

```haskell
module Wonderful where
```
While a filename would be `Wonderful.hs`.

My interpretation of the Haddock (I had to go there), was that `loadModule`'s arguments are `filepath` and `filename`. The documentation doesn't mention those. Instead it referts to _namespaces_, which could be either in the path (`very/uber/wonderful`) or the filename,(`very.uber.wonderful.theta`) or in a string within the `wonderful.theta` file, such as, say,

```
language-version: 1.0.0
avro-version: 1.0.0

---

type Very.Uber.Wonderful = Wow | Eww
```

By the way, did you describe the `---` anywhere? What is it for?

**Comment by TikhonJelvis** [comment]
Yeah, I can see how that's confusing. The idea is similar to how it works in Java: the load path tells us where to look in the filesystem as a whole, then, on top of that, there is a convention that namespaces get turned into directories and the module file ends in `.theta` (eg `com.target.foo` would be located in `com/target/foo.theta` somewhere in the load path).

`loadModule` takes the load path (which tells it where to look) and the module *name*; it then follows the convention I mentioned to find the actual `*.theta` file that defines the module with the given name.

This seems pretty similar to how it works in Haskell except that Haskell doesn't have the concept of a "load path" and instead needs modules to be explicitly specified in a `.cabal` file.

If that makes sense, how would you suggest rewording the documentation?

> By the way, did you describe the --- anywhere? What is it for?

Oh, that's just a syntactic separator between the metadata and the rest of the definitions. Just realized it isn't shown anywhere in the user guide, so I'll update that too.

**Comment by pasunboneleve** [comment]
> If that makes sense, how would you suggest rewording the documentation?

How about describing it as `filepath` and `filename`, and then explaining how this will then be reused as namespaces?

Also, at first it wasn't clear to me `type Wonderful = Aww | Eww` was going to work at all! Maybe a mention that the usual Haskell constructs would work as usual would be granted? Because my first reaction reading the documentation was _WHERE ARE THE ENUMS?!?_

BTW, great work!
And ... I'm curious. How does this library handle TH compilation time? I'm having slow compilation using Data.Avro, that's the motivation for looking around.

**Comment by TikhonJelvis** [comment]
I added a couple of paragraphs and examples to the User Guide section about `loadModule`; hopefully it makes more sense now.

> I'm curious. How does this library handle TH compilation time?

I didn't do anything special here. We have some medium-sized uses of Theta at Target and I don't recall compile times being a big problem. One thing that might help is that we can compile different Theta modules *in different Haskell modules*, which gives us incremental compilation for the TH-generated types—if the TH gets slow, at least you won't need to recompile the whole thing every time any bit of Theta changes.

The way importing *in Haskell* works is a bit janky and unintuitive, but I couldn't figure out a better solution given Template Haskell's limitations. Once I got used to the behavior, though, it was pretty natural to keep Theta schemas organized into separate Haskell modules and things seemed to work pretty well.

**Comment by TikhonJelvis** [comment]
Looks like the importing stuff isn't even documented in the User Guide right now :/. I'll probably add that later, but here's a quick overview:

Let's say we have two modules in `specs/a.theta` and `specs/b.theta`, where `b.theta` imports `a` (`import b`). If we just had a single file for `b`

``` haskell
module B where

import Theta.Target.Haskell (loadModule)

loadModule "specs" "b"
```

then it would always generate definitions for both `a.theta` *and* `b.theta`.

Instead, we can write two Haskell modules (`A.hs` and `B.hs`) and import `A` from `B`:

`A.hs`:

``` haskell
module A where

import Theta.Target.Haskell (loadModule)

loadModule "specs" "a"
```

`B.hs`:

``` haskell
module B where

import Theta.Target.Haskell (loadModule)

import A

loadModule "specs" "b"
```

As long as `A` is imported *unqualified*, `B` will only generate TH types for `b.theta`, and those types will reference the ones imported from `A`. If we left the `import A` out (or imported it qualified), it would generate new Haskell types for `a.theta` that would be incompatible with the ones from `A.hs`.

The fact that we have to import `A` unqualified points at how this is a bit of a hack under the hood—when `a` was loaded into TH, it generated a top-level value name `theta'a`; then, when we generate TH types for `b`, it looks at whether there is a value named `theta'a` already in scope and, if there is, doesn't generate new types for `a`.

Overall, I personally found this behavior to be really useful and straightforward, but I had *a lot* of trouble explaining it to other people so I realized it wasn't entirely intuitive. Unfortunately, I don't think there's any way to do a better job with TH because TH does not have any real way to do things across multiple modules (and also has no way I could find to deal with qualified imports).

**Comment by TikhonJelvis** [comment]
Oh, about enums:

```
type Wonderful = Aww | Eww
```

doesn't actually compile to an Avro enum—it's treated just like any other data type and compiles to a union of records. This was a conscious design decision, because it means that you could add another constructor *with* fields to the reader schema's version of `Wonderful` but still be backwards compatible with writer schemas that didn't have that field. That said, direct support for backwards compatibility is still pretty limited in Theta; it's on my to-do list, but it'll take some extra implementation and still has some open design questions.

I remember adding an explicit enum feature to Theta so that you could write 

```
enum Wonderful = Aww | Eww
```

and have that compile to an enum in Avro as well. However, it looks like I might have done that for the internal version of the library but not the open sourced version, or maybe it got lost in the open sourcing process somewhere :/. Either way, porting that feature over to the open source library should not be difficult.

**Comment by pasunboneleve** [comment]
> That said, direct support for backwards compatibility is still pretty limited in Theta; it's on my to-do list, but it'll take some extra implementation and still has some open design questions.

Well the selling point for Avro is forwards/backwards compatibility.

> Either way, porting that feature over to the open source library should not be difficult.

Well the point of using **theta-idl** is to be able to cross language boundaries. If that doesn't work across languages, that would be a problem.

**Comment by pasunboneleve** [comment]
Do you ever have issues with **haskell-language-server** and this library? We're having TH issues, it breaks LSP.

**Comment by TikhonJelvis** [comment]
Yeah, backwards compatibility is definitely important. We weren't using it heavily, and the one or two cases where we needed something non-trivial, people could use the generated Avro schemas directly without the TH-generated types. In general, being able to use raw Avro as an "escape hatch" was handy as Theta itself was being built since there was always a workaround for missing features or languages that didn't have direct Theta support yet. 

I had started working on something for backwards compatible parsing in Haskell just as I switched teams and put the project on hold, but I can probably resurrect that reasonably easily.

> Well the point of using theta-idl is to be able to cross language boundaries. If that doesn't work across languages, that would be a problem.

Enum support should work fine across different languages, it's just a matter of adding it here. That said, the way Theta is designed is that a Theta record always maps to valid Avro, but not every possible Avro schema can be reflected in Theta. This can be a definite limitation if you're getting Avro schemas from somewhere else, but I don't think it's possible to cover all of Avro while keeping the generated types and API nice to use in Haskell or Rust.

> Do you ever have issues with haskell-language-server and this library? We're having TH issues, it breaks LSP.

I've never tried it with LSP. Honestly, I still haven't used LSP much for anything in Haskell. I think there's some way to generate the TH definitions and save them to a file which might be a workaround, but I've never tried that myself.
